core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala

```diff
@@ -107,25 +107,25 @@ private[mesos] trait MesosSchedulerUtils extends Logging {
         return
       }
 
-      new Thread(Utils.getFormattedClassName(this) + "-mesos-driver") {
+      val t = new Thread(Utils.getFormattedClassName(this) + "-mesos-driver") {
         setDaemon(true)
 
         override def run() {
           mesosDriver = newDriver
-          try {
-            val ret = mesosDriver.run()
-            logInfo("driver.run() returned with code " + ret)
-            if (ret != null && ret.equals(Status.DRIVER_ABORTED)) {
-              System.exit(1)
-            }
-          } catch {
-            case e: Exception => {
-              logError("driver.run() failed", e)
-              System.exit(1)
-            }
+          val ret = mesosDriver.run()
+          logInfo("driver.run() returned with code " + ret)
+          if (ret != null && ret.equals(Status.DRIVER_ABORTED)) {
+            throw new SparkException("Error starting driver, DRIVER_ABORTED")
           }
         }
-      }.start()
+      }
+
+      t.start()
+      t.setUncaughtExceptionHandler(new Thread.UncaughtExceptionHandler {
```

Comment:
I'd move this to just below `setDaemon(true)` (containment) and before `t.start()` since it could fail before this line is executed.